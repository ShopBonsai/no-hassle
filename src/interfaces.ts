import * as Joi from 'joi';
import { RequestHandler } from 'express';
import {
  ComponentsObject,
  ExternalDocumentationObject,
  InfoObject,
  PathsObject,
  SecurityRequirementObject,
  ServerObject,
  TagObject,
} from 'openapi3-ts';

import { HttpMethod } from './constants';

export interface IRouteResult {
  get: (path: string, options: IOptions, ...args: RequestHandler[]) => IRouteResult;
  post: (path: string, options: IOptions, ...args: RequestHandler[]) => IRouteResult;
  put: (path: string, options: IOptions, ...args: RequestHandler[]) => IRouteResult;
  patch: (path: string, options: IOptions, ...args: RequestHandler[]) => IRouteResult;
  delete: (path: string, options: IOptions, ...args: RequestHandler[]) => IRouteResult;
  template: (template: ITemplateRoute, ...args: RequestHandler[]) => IRouteResult;
}

export interface ISwaggerDefinition {
  openapi: string;
  info: InfoObject;
  servers?: ServerObject[];
  paths: PathsObject;
  components?: ComponentsObject;
  security?: SecurityRequirementObject[];
  tags?: TagObject[];
  externalDocs?: ExternalDocumentationObject;
}

// Default options passed along when initiating Swagger in project
export interface ISwaggerOptions {
  title?: string;
  description?: string;
  servers?: ServerObject[];
  apiVersion?: '3.0';
  version?: string;
  auth?: 'basic';
  termsOfService?: string;
}

export interface IInput {
  body?: ISchema;
  query?: ISchema | object;
  params?: ISchema | object;
}

export type IOutput = {
  [key: number]: ISchema | { [key: string]: ISchema };
};

export interface IDocsOptions {
  input?: IInput;
  output?: IOutput;
  schemaOptions?: object;
  tags?: string[];
  description?: string;
  summary?: string;
}

export interface ITemplateRoute extends IDocsOptions {
  method: HttpMethod;
  path: string;
}

export interface IExecOptions extends IDocsOptions {
  docs?: boolean;
  method: HttpMethod;
  path: string;
  prefix?: string;
}

export interface IOptions extends IDocsOptions {
  docs?: boolean;
}

export interface IValidateResponse {
  isValid: boolean;
  errors?: Joi.ValidationErrorItem[];
}

export type ISchema = Joi.AnySchema;
