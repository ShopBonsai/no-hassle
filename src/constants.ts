export enum HttpMethod {
  Get = 'get',
  Post = 'post',
  Put = 'put',
  Patch = 'patch',
  Delete = 'delete',
}
