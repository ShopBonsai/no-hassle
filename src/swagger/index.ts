import { IDocsOptions, ISwaggerOptions, ISwaggerDefinition } from '../interfaces';
import { cleanPath } from '../lib/utils';
import { baseDefinition } from './baseDefinition';
import { getParameters } from './parameter';
import { getResponses } from './response';
import { getAuthentication } from './auth';
import { HttpMethod } from '../constants';

enum SwaggerDefaultConfig {
  VERSION = '2.0',
  HOST = '',
}

// TODO: move expectation to consumer
const globalSwagger: ISwaggerDefinition = {
  ...baseDefinition({
    title: 'Default swagger title',
    description: 'The API is documented here',
    host: 'localhost:3000',
    basePath: '/',
    version: '1.0',
    apiVersion: '2.0',
    contact: { email: 'developers@shopbonsai.com' },
  }),
};

export const updateSwagger = (key: string, values: object) =>
  Object.assign(globalSwagger[key], values);

export const generateSwagger = (path: string, method: HttpMethod, options: IDocsOptions) => {
  const {
    input,
    output,
    contentTypes = ['application/json'],
    summary,
    description,
    tags = [],
  } = options;

  const cleanedPath = cleanPath(path);

  const result = {
    [method]: {
      tags,
      produces: contentTypes,
      parameters: getParameters(globalSwagger, input),
      responses: getResponses(globalSwagger, output),
      // optional
      ...(summary && { summary }),
      ...(description && { description }),
    },
  };
  // If path already exists (other method for example)
  if (globalSwagger.paths.hasOwnProperty(cleanedPath)) {
    Object.assign(globalSwagger.paths[cleanedPath], result);
  } else {
    updateSwagger('paths', { [cleanedPath]: result });
  }
};

export const getSwagger = (options: ISwaggerOptions = {}): ISwaggerDefinition => {
  const {
    apiVersion = SwaggerDefaultConfig.VERSION,
    host = SwaggerDefaultConfig.HOST,
    auth,
    ...otherOptions
  } = options;

  const authentication = getAuthentication(auth);
  const result: ISwaggerDefinition = {
    ...globalSwagger,
    ...authentication,
    host,
    info: {
      ...globalSwagger.info,
      ...otherOptions,
    },
  };

  return result;
};
