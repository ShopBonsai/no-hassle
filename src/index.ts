import { Application, RequestHandler } from 'express';
import { isCelebrate } from 'celebrate';
import { getSwagger, generateSwagger } from './swagger';
import { validateSchema, validateValue } from './lib/validator';
import { IOptions, IRouteResult, IExecOptions, ITemplateRoute } from './interfaces';
import { HttpMethod } from './constants';

const addRoute = (
  app: Application,
  options: IExecOptions,
  ...args: RequestHandler[]
): IRouteResult => {
  const {
    method,
    path: rawPath,
    prefix = '',
    input = null,
    schemaOptions = {},
    docs = true,
  } = options;
  const path = rawPath === '/' ? prefix : `${prefix}${rawPath}`;

  // TODO: Add method overloading -> no options provided -> basic result

  // Autogenerate documentation based on the schema
  if (docs) {
    generateSwagger(path, method, options);
  }

  // Add express route (with optional validation)
  input && Object.keys(input).length > 0
    ? app[method](path, validateSchema(input, schemaOptions), ...args)
    : app[method](path, args);

  // Return function instance with same router for currying
  return router.use(app, prefix);
};

export const router = {
  use: (app: Application, prefix: string = ''): IRouteResult => ({
    get: (path, options, ...args) =>
      addRoute(app, { ...options, path, prefix, method: HttpMethod.Get }, ...args),
    post: (path, options, ...args) =>
      addRoute(app, { ...options, path, prefix, method: HttpMethod.Post }, ...args),
    put: (path, options, ...args) =>
      addRoute(app, { ...options, path, prefix, method: HttpMethod.Put }, ...args),
    patch: (path, options, ...args) =>
      addRoute(app, { ...options, path, prefix, method: HttpMethod.Patch }, ...args),
    delete: (path, options, ...args) =>
      addRoute(app, { ...options, path, prefix, method: HttpMethod.Delete }, ...args),
    template: ({ path, method, ...otherOpts }, ...args) =>
      addRoute(app, { ...otherOpts, path, prefix, method }, ...args),
  }),
};

export const isValidationError = (error: object): boolean => isCelebrate(error);

export * from './swagger/errors';
export * from './constants';
export * from './interfaces';
export { IOptions, ITemplateRoute, getSwagger, validateValue, validateSchema };
