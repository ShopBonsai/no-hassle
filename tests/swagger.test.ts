import * as express from 'express';
import * as Joi from 'joi';

import { errorSchema, getSwagger, router } from '../src';
import { productSchema } from './mocks/product';

describe('Swagger generation', () => {
  describe('getSwagger', () => {
    it('Should return autogenerated default Swagger definition', () => {
      const swagger = getSwagger({
        version: '1.0.0',
        title: 'Test API',
        description: 'Test API description',
        servers: [{ url: 'https://myapi.org' }],
      });

      expect(swagger).toMatchSnapshot();
    });
  });

  describe('addRoute', () => {
    const app = express();

    it('Should succesfully generate swagger route', () => {
      router.use(app, '/products').post('/', {
        description: 'Create new product',
        tags: ['Products'],
        input: {
          body: productSchema,
        },
        output: {
          200: productSchema, // Direct schema - not wrapped as object
          400: {
            data: productSchema,
            error: errorSchema, // Single error
          },
          401: {
            data: productSchema,
            errors: Joi.array().items(errorSchema), // Array of errors
          },
        },
      });

      const swagger = getSwagger({
        version: '1.0.0',
        title: 'Test API',
        description: 'Test API description',
        servers: [{ url: 'https://myapi.org' }],
      });

      expect(swagger).toMatchSnapshot();
    });
  });
});
