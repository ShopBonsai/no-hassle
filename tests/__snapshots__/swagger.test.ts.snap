// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Swagger generation addRoute Should succesfully generate swagger route 1`] = `
Object {
  "components": Object {
    "schemas": Object {
      "8s2h35vjiResult": Object {
        "additionalProperties": false,
        "properties": Object {
          "detail": Object {
            "example": "Fatal issue",
            "nullable": true,
            "type": "string",
          },
          "status": Object {
            "example": 500,
            "format": "float",
            "type": "number",
          },
          "title": Object {
            "example": "Unknown error",
            "type": "string",
          },
        },
        "required": Array [
          "status",
          "detail",
          "title",
        ],
        "type": "object",
      },
      "BadRequestError": Object {
        "additionalProperties": false,
        "description": "Provided input is invalid",
        "properties": Object {
          "errors": Object {
            "items": Object {
              "additionalProperties": false,
              "example": Object {
                "detail": "",
                "status": 400,
                "title": "Bad input request",
              },
              "properties": Object {
                "detail": Object {
                  "example": "Fatal issue",
                  "nullable": true,
                  "type": "string",
                },
                "status": Object {
                  "example": 500,
                  "format": "float",
                  "type": "number",
                },
                "title": Object {
                  "example": "Unknown error",
                  "type": "string",
                },
              },
              "required": Array [
                "status",
                "detail",
                "title",
              ],
              "type": "object",
            },
            "type": "array",
          },
        },
        "type": "object",
      },
      "NotFoundError": Object {
        "additionalProperties": false,
        "description": "Not Found",
        "properties": Object {
          "errors": Object {
            "items": Object {
              "additionalProperties": false,
              "example": Object {
                "detail": "",
                "status": 404,
                "title": "Not Found",
              },
              "properties": Object {
                "detail": Object {
                  "example": "Fatal issue",
                  "nullable": true,
                  "type": "string",
                },
                "status": Object {
                  "example": 500,
                  "format": "float",
                  "type": "number",
                },
                "title": Object {
                  "example": "Unknown error",
                  "type": "string",
                },
              },
              "required": Array [
                "status",
                "detail",
                "title",
              ],
              "type": "object",
            },
            "type": "array",
          },
        },
        "type": "object",
      },
      "UnauthorizedError": Object {
        "additionalProperties": false,
        "description": "Authentication information is missing or invalid",
        "properties": Object {
          "errors": Object {
            "items": Object {
              "additionalProperties": false,
              "example": Object {
                "detail": "",
                "status": 401,
                "title": "Authentication information is missing or invalid",
              },
              "properties": Object {
                "detail": Object {
                  "example": "Fatal issue",
                  "nullable": true,
                  "type": "string",
                },
                "status": Object {
                  "example": 500,
                  "format": "float",
                  "type": "number",
                },
                "title": Object {
                  "example": "Unknown error",
                  "type": "string",
                },
              },
              "required": Array [
                "status",
                "detail",
                "title",
              ],
              "type": "object",
            },
            "type": "array",
          },
        },
        "type": "object",
      },
      "UnknownError": Object {
        "additionalProperties": false,
        "description": "Unknown error",
        "properties": Object {
          "errors": Object {
            "items": Object {
              "additionalProperties": false,
              "example": Object {
                "detail": "",
                "status": 500,
                "title": "Unknown error",
              },
              "properties": Object {
                "detail": Object {
                  "example": "Fatal issue",
                  "nullable": true,
                  "type": "string",
                },
                "status": Object {
                  "example": 500,
                  "format": "float",
                  "type": "number",
                },
                "title": Object {
                  "example": "Unknown error",
                  "type": "string",
                },
              },
              "required": Array [
                "status",
                "detail",
                "title",
              ],
              "type": "object",
            },
            "type": "array",
          },
        },
        "type": "object",
      },
      "kr0jt3p1wResult": Object {
        "items": Object {
          "additionalProperties": false,
          "properties": Object {
            "detail": Object {
              "example": "Fatal issue",
              "nullable": true,
              "type": "string",
            },
            "status": Object {
              "example": 500,
              "format": "float",
              "type": "number",
            },
            "title": Object {
              "example": "Unknown error",
              "type": "string",
            },
          },
          "required": Array [
            "status",
            "detail",
            "title",
          ],
          "type": "object",
        },
        "type": "array",
      },
      "universalBrandInput": Object {
        "additionalProperties": false,
        "properties": Object {
          "country": Object {
            "description": "Country code.",
            "example": "US",
            "type": "string",
          },
          "description": Object {
            "description": "Description of the brand.",
            "example": "This brand is amazing.",
            "nullable": true,
            "type": "string",
          },
          "logo": Object {
            "description": "Brand logo image.",
            "example": "http://images.ctfassets.net/hhq3zxylfx2l/H5laGR7EAeMQEeYqEmyGm/06a443581102ea4e52ea8294f8723790/Bonsai_image.jpg",
            "nullable": true,
            "type": "string",
          },
          "name": Object {
            "description": "Brand name.",
            "example": "Virgin Club",
            "type": "string",
          },
          "public_id": Object {
            "description": "Brand public identifier.",
            "example": "1234asfsdfhwadasd2v",
            "type": "string",
          },
        },
        "required": Array [
          "public_id",
          "name",
        ],
        "type": "object",
      },
      "universalBrandResult": Object {
        "additionalProperties": false,
        "properties": Object {
          "country": Object {
            "description": "Country code.",
            "example": "US",
            "type": "string",
          },
          "description": Object {
            "description": "Description of the brand.",
            "example": "This brand is amazing.",
            "nullable": true,
            "type": "string",
          },
          "logo": Object {
            "description": "Brand logo image.",
            "example": "http://images.ctfassets.net/hhq3zxylfx2l/H5laGR7EAeMQEeYqEmyGm/06a443581102ea4e52ea8294f8723790/Bonsai_image.jpg",
            "nullable": true,
            "type": "string",
          },
          "name": Object {
            "description": "Brand name.",
            "example": "Virgin Club",
            "type": "string",
          },
          "public_id": Object {
            "description": "Brand public identifier.",
            "example": "1234asfsdfhwadasd2v",
            "type": "string",
          },
        },
        "required": Array [
          "public_id",
          "name",
        ],
        "type": "object",
      },
    },
  },
  "info": Object {
    "contact": Object {
      "email": "developers@shopbonsai.com",
    },
    "description": "Test API description",
    "title": "Test API",
    "version": "1.0.0",
  },
  "openapi": "3.0.1",
  "paths": Object {
    "/products": Object {
      "post": Object {
        "description": "Create new product",
        "parameters": Array [],
        "requestBody": Object {
          "content": Object {
            "application/json": Object {
              "schema": Object {
                "$ref": "#/components/schemas/universalBrandInput",
              },
            },
          },
          "required": true,
        },
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "$ref": "#/components/schemas/universalBrandResult",
                },
              },
            },
            "description": "Success",
          },
          "400": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "properties": Object {
                    "data": Object {
                      "$ref": "#/components/schemas/universalBrandResult",
                    },
                    "error": Object {
                      "$ref": "#/components/schemas/8s2h35vjiResult",
                    },
                  },
                  "type": "object",
                },
              },
            },
            "description": "400 - Bad request",
          },
          "401": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "properties": Object {
                    "data": Object {
                      "$ref": "#/components/schemas/universalBrandResult",
                    },
                    "errors": Object {
                      "$ref": "#/components/schemas/kr0jt3p1wResult",
                    },
                  },
                  "type": "object",
                },
              },
            },
            "description": "401 - Unauthorized",
          },
          "404": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "$ref": "#/components/schemas/NotFoundError",
                },
              },
            },
            "description": "404 - Not found",
          },
          "405": Object {
            "content": Object {},
            "description": "405 - Validation exception",
          },
          "5XX": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "$ref": "#/components/schemas/UnknownError",
                },
              },
            },
            "description": "500 - Unknown error",
          },
        },
        "tags": Array [
          "Products",
        ],
      },
    },
  },
  "servers": Array [
    Object {
      "url": "https://myapi.org",
    },
  ],
}
`;

exports[`Swagger generation getSwagger Should return autogenerated default Swagger definition 1`] = `
Object {
  "components": Object {
    "schemas": Object {
      "BadRequestError": Object {
        "additionalProperties": false,
        "description": "Provided input is invalid",
        "properties": Object {
          "errors": Object {
            "items": Object {
              "additionalProperties": false,
              "example": Object {
                "detail": "",
                "status": 400,
                "title": "Bad input request",
              },
              "properties": Object {
                "detail": Object {
                  "example": "Fatal issue",
                  "nullable": true,
                  "type": "string",
                },
                "status": Object {
                  "example": 500,
                  "format": "float",
                  "type": "number",
                },
                "title": Object {
                  "example": "Unknown error",
                  "type": "string",
                },
              },
              "required": Array [
                "status",
                "detail",
                "title",
              ],
              "type": "object",
            },
            "type": "array",
          },
        },
        "type": "object",
      },
      "NotFoundError": Object {
        "additionalProperties": false,
        "description": "Not Found",
        "properties": Object {
          "errors": Object {
            "items": Object {
              "additionalProperties": false,
              "example": Object {
                "detail": "",
                "status": 404,
                "title": "Not Found",
              },
              "properties": Object {
                "detail": Object {
                  "example": "Fatal issue",
                  "nullable": true,
                  "type": "string",
                },
                "status": Object {
                  "example": 500,
                  "format": "float",
                  "type": "number",
                },
                "title": Object {
                  "example": "Unknown error",
                  "type": "string",
                },
              },
              "required": Array [
                "status",
                "detail",
                "title",
              ],
              "type": "object",
            },
            "type": "array",
          },
        },
        "type": "object",
      },
      "UnauthorizedError": Object {
        "additionalProperties": false,
        "description": "Authentication information is missing or invalid",
        "properties": Object {
          "errors": Object {
            "items": Object {
              "additionalProperties": false,
              "example": Object {
                "detail": "",
                "status": 401,
                "title": "Authentication information is missing or invalid",
              },
              "properties": Object {
                "detail": Object {
                  "example": "Fatal issue",
                  "nullable": true,
                  "type": "string",
                },
                "status": Object {
                  "example": 500,
                  "format": "float",
                  "type": "number",
                },
                "title": Object {
                  "example": "Unknown error",
                  "type": "string",
                },
              },
              "required": Array [
                "status",
                "detail",
                "title",
              ],
              "type": "object",
            },
            "type": "array",
          },
        },
        "type": "object",
      },
      "UnknownError": Object {
        "additionalProperties": false,
        "description": "Unknown error",
        "properties": Object {
          "errors": Object {
            "items": Object {
              "additionalProperties": false,
              "example": Object {
                "detail": "",
                "status": 500,
                "title": "Unknown error",
              },
              "properties": Object {
                "detail": Object {
                  "example": "Fatal issue",
                  "nullable": true,
                  "type": "string",
                },
                "status": Object {
                  "example": 500,
                  "format": "float",
                  "type": "number",
                },
                "title": Object {
                  "example": "Unknown error",
                  "type": "string",
                },
              },
              "required": Array [
                "status",
                "detail",
                "title",
              ],
              "type": "object",
            },
            "type": "array",
          },
        },
        "type": "object",
      },
    },
  },
  "info": Object {
    "contact": Object {
      "email": "developers@shopbonsai.com",
    },
    "description": "Test API description",
    "title": "Test API",
    "version": "1.0.0",
  },
  "openapi": "3.0.1",
  "paths": Object {},
  "servers": Array [
    Object {
      "url": "https://myapi.org",
    },
  ],
}
`;
